compiler/next/include/chpl/resolution/resolution-types.h

```diff
@@ -35,105 +39,553 @@ namespace resolution {
+  // be reused given the PoiInfo for a not-yet-resolved function in 'check'.
+  bool canReuse(const PoiInfo& check) const;
+
+  // return true if one of the PoiInfos is a resolved function that
+  // can be reused given PoiInfo for a not-yet-resolved function.
+  static bool reuseEquals(const PoiInfo& a, const PoiInfo& b) {
+    if (a.resolved && !b.resolved) {
+      return a.canReuse(b);
+    }
+    if (b.resolved && !a.resolved) {
+      return b.canReuse(a);
+    }
+    return updateEquals(a, b);
+  }
+
+  // hashing a PoiInfo gives 0 always
+  // (instead we rely on == in the hashtable so that we can
+  //  apply canReuse to figure out if an instantiation can be reused).
+  size_t hash() const {
+    return 0;
+  }
+  // == and != for the hashtable
+  bool operator==(const PoiInfo& other) const {
+    return PoiInfo::reuseEquals(*this, other);
+  }
+  bool operator!=(const PoiInfo& other) const {
+    return !(*this == other);
+  }
 };
 
+// TODO: should this actually be types::FunctionType?
 struct TypedFnSignature {
-  UniqueString name;
-  bool isMethod; // in that case, formals[0] is the receiver
-  std::vector<TypedFnSignatureFormal> formals;
+  typedef enum {
+    WHERE_NONE,  // no where clause
+    WHERE_TBD,   // where clause not resolved yet
+    WHERE_TRUE,  // where resulted in true
+    WHERE_FALSE, // where resulted in false
+  } WhereClauseResult;
+
+  // What is the untyped function signature?
+  const UntypedFnSignature* untypedSignature;
+  // What is the type of each of the formals?
+  std::vector<types::QualifiedType> formalTypes;
+  // If there was a where clause, what was the result of evaluating it?
+  WhereClauseResult whereClauseResult = WHERE_TBD;
+
+  // Are any of the formals generic or unknown at this point?
+  bool needsInstantiation = true;
+
+  // Is this TypedFnSignature representing an instantiation?
+  // If so, what is the generic TypedFnSignature that was instantiated?
+  const TypedFnSignature* instantiatedFrom = nullptr;
+
+  // Is this for an inner Function? If so, what is the parent
+  // function signature?
+  const TypedFnSignature* parentFn = nullptr;
+
+  // TODO: This could include a substitutions map, if we need it.
+  // The formalTypes above might be enough, though.
+
+  bool operator==(const TypedFnSignature& other) const {
+    return untypedSignature == other.untypedSignature &&
+           formalTypes == other.formalTypes &&
+           whereClauseResult == other.whereClauseResult &&
+           needsInstantiation == other.needsInstantiation &&
+           instantiatedFrom == other.instantiatedFrom &&
+           parentFn == other.parentFn;
+  }
+  bool operator!=(const TypedFnSignature& other) const {
+    return !(*this == other);
+  }
+
+  std::string toString() const;
+
+  const ID& functionId() const {
+    return untypedSignature->functionId;
+  }
+};
+
+class MostSpecificCandidates {
+ public:
+  typedef enum {
+    REF = 0,
+    CONST_REF = 1,
+    VALUE = 2,
+    N = 3,
```

Comment:
Yes. But I don't see it as a particular problem. I could imagine that we might be sad that a `switch` over this enum would have to handle `N` but it's not an enum I expect to ever `switch` over. So I don't see any reason not to have this N here. Can you say more about what your concern is?

Reply:
I'm mostly just curious if there's another way to handle this that feels less like a work around.  I know it's pretty common in C++ to have an `end()` value you can compare against, so I'm more wondering if there's a different way they implement that which would be considered more standard.  Right now, this seems like a workaround and that's my main objection to it - if there's a cleaner way that doesn't add unnecessary clutter, I'd like us to use it.  If there isn't, that's fine (though I would like it to have a name like `NUM_INTENTS` to make it explicitly clear that this is not a value we expect to use except for in the case of iterating over it, and maybe a comment about why it is different than the other enum constants)

Reply:
I can't think of any way that not having an `N` in some form would make this simpler. I don't think of it as a workaround. We must know, at compile time, the length of the array, for it to be a C array. It would add unnecessary overhead to use e.g. a `std::vector` here. I did implement an `end` but it uses `N`. I could use `VALUE+1` everywhere instead of `N` but that seems strictly worse to me.

That said, it would not bother me to make `N` private or something. But as it is, if the `Intent` enum is updated, it should be easy to update `N`. 

I am happy to rename `N` to `NUM_INTENTS` and to add a comment.